### ISSUE TITLE
ndsize_t is not a valid data type

### ISSUE THREAD

**Issue by jgrewe** [G-Node/nix#486]
error: ‘value’ is not a member of ‘nix::to_data_type<long long unsigned int>’
     std::cerr <<  nix::to_data_type<ndsize_t>::value << std::endl;
I tried the following code

````
    std::cerr <<  nix::to_data_type<ndsize_t>::is_valid << std::endl;
    std::cerr <<  nix::to_data_type<ndsize_t>::value << std::endl;
````

Which does not compile with the error message:
````
     error: ‘value’ is not a member of ‘nix::to_data_type<long long unsigned int>’
     std::cerr <<  nix::to_data_type<ndsize_t>::value << std::endl;
````

The first line only gives a false.
Apparently this does not affect Mac of Win systems. My system: ubuntu 14.04, g++ version 4.8.2

update: 
sizeof ndsize_t and uint64_t match, both are 8 byte, as they should...

**Comment by gicmo** [comment]
I stumbled upon a similar bug today in [nix-mx](https://www.github.com/G-Node/nix-mx) but with `unsigned long`. 
It seems to occur on platforms where `long long unsigned` and `long unsigned` are the same size (see below), thus only one of them can be covered by our `nix::to_data_type<uint64_t>` template specialisation.

On current 64bit OSX 10.10.3 with clang [Apple LLVM version 6.1.0 (clang-602.0.53) (based on LLVM 3.6.0svn]:
 ```
% cat << EOF | clang -x c++ -o sizetest -std=c++11 -lc++ - && ./sizetest                                                            
#include <iostream>

int main(int argc, char **argv) {
        std::cerr << "UL " << sizeof(long unsigned int) << std::endl;
        std::cerr << "ULL " << sizeof(long long unsigned int) << std::endl;

        return 0;
}
EOF
UL 8
ULL 8
```

**Comment by gicmo** [comment]
After giving this some though maybe the best way to solve these issues is by redesigning the `nix::DataType` enums in such a way that the individual items encode their properties like the cube [cube_dtype_t](https://github.com/gicmo/cube/blob/master/lib/cube_array.h#L15). Then one should be able to come up with an `to_data_type<T>` that returns the appropriate enum item by just constructing it (setting the individual bits).

**Comment by gicmo** [comment]
This is currently breaking nix-mx on OSX:

```
In file included from /Users/gicmo/Coding/G-Node/nix-mx/nix_mx.cc:13:
In file included from /Users/gicmo/Coding/G-Node/nix-mx/src/utils/arguments.h:8:
/Users/gicmo/Coding/G-Node/nix-mx/src/utils/mknix.h:20:47: error: no member
      named 'value' in 'nix::to_data_type<unsigned long>'
    check_arg_type(arr, nix::to_data_type<T>::value);
                        ~~~~~~~~~~~~~~~~~~~~~~^
/Users/gicmo/Coding/G-Node/nix-mx/src/utils/arguments.h:64:16: note: in
      instantiation of function template specialization 'mx_to_num<unsigned
      long>' requested here
        return mx_to_num<T>(array[pos]);
```

**Comment by gicmo** [comment]
PR #535 did indeed fix issue G-Node/nix-mx#12 on OSX. I am confident that also the original `ndsize_t` issue is gone, but maybe @jgrewe can confirm.

**Comment by jgrewe** [comment]
indeed, the issue is fixed and hence I dare to close it :) Thanks
